### PR TITLE
GRPC server

### PR DESCRIPTION
**Pull request by astradzhao**

### CODE REVIEW

File: src/subcommands/subcommand.go [hunk 1]
```diff
@@ -186,20 +218,31 @@ func executeAdd(args []string) {
...
 		IndexPath:  *indexPath,
 		DataType:   *dataType,
 		SourceType: *sourceType,
 		MakeCopy:   *makeLocalCopy,
-		DataURIs:   cmd.Args(),
+		DataUris:   dataUris,
 	}
 
-	subcommands.Add(addArgs)
+	res, err := grpcclient.Add(addArgs)
```
Comment (fredxfred) [inline src/subcommands/subcommand.go]: Refactor Add in subcommands/add.go to take AddRequest (the proto, not the struct) as an argument and instead simply call Add without having created the grpcclient.

i.e.

Add(a MyAddProto)

If the actual grpc method is called Add you may wish to rename Add in subcommand.go to AddInternal or something.

Basically we should leave subcommand.go almost exactly the same as it is now, and just pass Protobufs to the subcommands instead of the (very similar/same) go structs we use currently on main.

File: src/grpcclient/grpcclient.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	pb "accretional.com/semantifly/proto/accretional.com/semantifly/proto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var client pb.SemantiflyClient
+var conn *grpc.ClientConn
+
+func Init() {
```
Comment (fredxfred) [inline src/grpcclient/grpcclient.go]: Either make this part of the constructor or return an err

You should always prefer to write init code in constructors rather than separate functions. The one exception is if you need to return something, since constructors can't do that and you often might want to return err types when dealing with eg network code (like this) which is likely going to fail sometimes. In this particular case I'd suggest putting it in the constructor because we don't really need to implement error handling for a failed network initialization yet

File: src/grpcclient/grpcclient.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	_, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// Create a gRPC client using the grpc.NewClient function
+	var err error
```
Comment (fredxfred) [inline src/grpcclient/grpcclient.go]: Don't declare var errors, just do ```conn, err := ...```

File: src/proto/semantifly.proto [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+service Semantifly {
+  rpc Add(AddRequest) returns (AddResponse) {}
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+  rpc Get(GetRequest) returns (GetResponse) {}
+  rpc Update(UpdateRequest) returns (UpdateResponse) {}
+  rpc LexicalSearch(LexicalSearchRequest) returns (LexicalSearchResponse) {}
+}
+
+message AddRequest {
+  string index_path = 1;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: remove index_path from all of these proto messages, this should be supplied at init for the grpc server via command line args. Using it here leaks implementation details to the client and also obviates most of the reason for running semantifly as a server (having the index already loaded into memory when serving eg a Get request)

File: src/proto/semantifly.proto [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  string data_type = 2;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: These should be enums just as they are in the Index proto. In fact you should probably refactor the index proto and this to share message types

File: src/proto/semantifly.proto [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  string source_type = 3;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: ditto make this an enum

File: src/proto/semantifly.proto [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  bool make_copy = 4;
+  repeated string data_uris = 5;
+}
+
+message FileContent {
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: where is this used?

File: src/proto/semantifly.proto [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+message LexicalSearchRequest {
+  string index_path = 1;
+  string search_term = 2;
+  int32 top_n = 3;
+}
+
+message LexicalSearchResponse {
+  bool success = 1;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: Don't use success fields in proto responses, grpc has builtin functions to check that

File: src/proto/semantifly.proto [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  repeated LexicalSearchResult results = 2;
+  string message = 3;
+}
+
+message LexicalSearchResult {
+  string name = 1;
+  float occurrences = 2;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: Why a float? In general you should never never never use floats unless writing scientific/ml/etc code that benefits from the FPU speedup of operating on floats

File: src/proto/semantifly.proto [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+message UpdateRequest {
+  string name = 1;
+  DataType data_type = 2;
+  SourceType source_type = 3;
+  bool update_copy = 4;
+  string data_uri = 5;
+}
+
+message UpdateResponse {
+  string message = 1;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: What kind of message? Error message probably right? So maybe call this (and other "messages" used for the same thing) error_message

File: src/proto/semantifly.proto [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  repeated LexicalSearchResult results = 1;
+  string message = 2;
+}
+
+message LexicalSearchResult {
+  string name = 1;
+  int32 occurrences = 2;
+}
+
+message Index {
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: Please move Index, ILE, and other types back to their original file and import them https://protobuf.dev/programming-guides/proto3/#importing

File: src/proto/semantifly.proto (hunk 10 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: message stores all messages in the console (e.g. output messages, error messages that still allow the function to continue, etc.) not necessarily only error messages

File: src/proto/semantifly.proto (hunk 7 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 8 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 9 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 6 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 5 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 4 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto (hunk 11 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+service Semantifly {
+  rpc Add(AddRequest) returns (AddResponse) {}
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+  rpc Get(GetRequest) returns (GetResponse) {}
+  rpc Update(UpdateRequest) returns (UpdateResponse) {}
+  rpc LexicalSearch(LexicalSearchRequest) returns (LexicalSearchResponse) {}
+}
+
+message AddRequest {
+  DataType data_type = 1;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: use ContentMetadata here and elsewhere

File: src/proto/semantifly.proto (hunk 10 above)
Comment (fredxfred) [reply src/proto/semantifly.proto]: We should not output any non-error/warning messages in this response. Please only include warnings and actual errors and change this to error_message

File: src/subcommands/server.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	pb "accretional.com/semantifly/proto/accretional.com/semantifly/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Server struct {
+	pb.UnimplementedSemantiflyServer
+}
+
+const serverIndexPath = "index"
```
Comment (fredxfred) [inline src/subcommands/server.go]: Why is this a const? Shouldn't it be plumbed down from the command line arguments

File: src/proto/semantifly.proto (hunk 10 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: done

File: src/proto/semantifly.proto (hunk 12 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/subcommands/server.go (hunk 13 above)
Comment (astradzhao) [reply src/subcommands/server.go]: fixed

File: src/proto/semantifly.proto (hunk 10 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/proto/semantifly.proto [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  bool make_copy = 2;
+}
+
+message AddResponse {
+  string error_message = 1;
+}
+
+message DeleteRequest {
+  bool delete_copy = 1;
+  repeated string data_uris = 2;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: Shouldn't these be names?

File: src/proto/semantifly.proto [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+message DeleteResponse {
+  string error_message = 1;
+}
+
+message GetRequest {
+  string name = 1;
+}
+
+message GetResponse {
```
Comment (fredxfred) [inline src/proto/semantifly.proto:38]: Add ContentMetadata to the GetResponse and make content optional. It could be possible that the content is stored locally to the client/authenticated and not accessible by the server, in which case we'd want to tell the client how to access the data even if the server itself cannot

File: src/proto/semantifly.proto [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+service Semantifly {
+  rpc Add(AddRequest) returns (AddResponse) {}
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+  rpc Get(GetRequest) returns (GetResponse) {}
+  rpc Update(UpdateRequest) returns (UpdateResponse) {}
+  rpc LexicalSearch(LexicalSearchRequest) returns (LexicalSearchResponse) {}
+}
+
+message AddRequest {
+  repeated ContentMetadata files_data = 1;
```
Comment (fredxfred) [inline src/proto/semantifly.proto]: It seems weird to have this be repeated but none of the other CRUD rpcs take repeated args, but let's not address this yet unless it's easy

Comment (fredxfred) [inline src/proto/semantifly.proto]: this field will not just contain files, so let's call it something like just "added_metadata"

File: src/proto/semantifly.proto (hunk 14 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: done

File: src/proto/semantifly.proto (hunk 15 above)
Comment (astradzhao) [reply src/proto/semantifly.proto:38]: fixed

File: src/proto/semantifly.proto (hunk 16 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/subcommands/server_test.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+		URI:        filepath.Join(testDir, "test_file1.txt"),
+	}
+
+	addArgs := &pb.AddRequest{
+		AddedMetadata: testFileData1,
+		MakeCopy:  true,
+	}
+
+	_, err := client.Add(addCtx, addArgs)
+	addCancel()
```
Comment (fredxfred) [inline src/subcommands/server_test.go]: why call this / why is this the only one that is not deferred?

File: src/subcommands/server_test.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	// Delete
+	delCtx, delCancel := context.WithTimeout(context.Background(), time.Second)
+	defer delCancel()
+
+	delReq := &pb.DeleteRequest{
+		DeleteCopy: true,
+		Names:   []string{filepath.Join(testDir, "test_file1.txt")},
+	}
+
+	_, err = client.Delete(delCtx, delReq)
```
Comment (fredxfred) [inline src/subcommands/server_test.go]: for this and other subcommands with an error message response, you should check that that field is empty

(also, add a test case that verifies that the error message field is non-empty when you deliberately trigger a bug. This should be in a new unit test)

File: src/subcommands/server_test.go [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+	getReq := &pb.GetRequest{
+		Name: filepath.Join(testDir, "test_file1.txt"),
+	}
+
+	getResp, err := client.Get(getCtx, getReq)
+	if err != nil {
+		t.Fatalf("Get failed: %v", err)
+	}
+	if *getResp.Content == "" {
```
Comment (fredxfred) [inline src/subcommands/server_test.go:140]: verify that the actual contents match with what you set them as instead of this

File: src/proto/semantifly.proto (hunk 16 above)
Comment (astradzhao) [reply src/proto/semantifly.proto]: fixed

File: src/subcommands/server_test.go (hunk 19 above)
Comment (astradzhao) [reply src/subcommands/server_test.go:140]: done

File: src/subcommands/server_test.go (hunk 18 above)
Comment (astradzhao) [reply src/subcommands/server_test.go]: done

File: src/subcommands/server_test.go (hunk 17 above)
Comment (astradzhao) [reply src/subcommands/server_test.go]: done

File: src/subcommands/server_test.go [hunk 20]
```diff
@@ -96,6 +97,32 @@ func TestServerCommands(t *testing.T) {
...
+		MakeCopy:      true,
+	}
+
+	_, err = client.Add(addCtx2, addArgs2)
+
+	if err.Error() == "" {
+		t.Fatalf("Failed to return error message when expected to")
+	}
+
+	expectedErrorMsg := "make_copy:true has already been added. Skipping without refresh."
```
Comment (fredxfred) [inline src/subcommands/server_test.go]: shouldn't this be a name or a URI?
